packages/intl-datetimeformat/src/core.ts

```diff
@@ -414,8 +426,11 @@ function initializeDateTimeFormat(
     if (matcher === 'basic') {
       bestFormat = basicFormatMatcher(opt, formats);
     } else {
-      opt.hour12 =
-        internalSlots.hourCycle === 'h11' || internalSlots.hourCycle === 'h12';
+      if (isTimeRelated(opt)) {
```

Comment:
@longlho to avoid nested looping I extracted a helper function, I tried to mimic the code style of the rest of this file, but let me know if you had some different approach in mind